tests/integration/test_charm.py

```diff
@@ -75,8 +81,130 @@ async def test_database_is_up_as_replica_set(ops_test: OpsTest, unit_id: int) ->
     client.close()
 
 
+async def test_all_units_in_same_replica_set(ops_test: OpsTest) -> None:
+    """Tests that all units in the application belong to the same replica set.
+
+    This test will be implemented by Raul as part of his onboarding task.
+
+    Hint: multiple ways to do this, here are two but there are probably
+    other ways to implement it:
+    - check that each unit has the same replica set name
+    - create a URI for all units for the expected replica set name and see if
+      you can connect
+    """
+    pass
+
+
+async def test_leader_is_primary_on_deployment(ops_test: OpsTest) -> None:
+    """Tests that right after deployment that the primary unit is the leader.
+    """
+    # grab leader unit
+    leader_unit = None
+    for unit in ops_test.model.applications[APP_NAME].units:
+        if await unit.is_leader_from_status():
+            leader_unit = unit
+
+    # verify that we have a leader
+    assert leader_unit is not None, "No unit is leader"
+
+    # connect to mongod
+    client = MongoClient(unit_uri(leader_unit.public_address))
+
+    # verify primary status
+    assert client.is_primary, "Leader is not primary"
+    client.close()
+
+
+async def test_exactly_one_primary(ops_test: OpsTest) -> None:
+    """Tests that there is exactly one primary in the deployed units.
+    """
+    try:
+        number_of_primaries = count_primaries(ops_test)
+    except RetryError:
+        number_of_primaries = 0
+
+    # check that exactly of the units is the leader
+    assert number_of_primaries == 1, "Expected one unit to be a primary: %s != 1" % (
+        number_of_primaries
+    )
+
+
+async def test_get_primary_action(ops_test: OpsTest) -> None:
+    """Tests that the action get-primary outputs the correct unit where the
+    primary replica is located.
+    """
+    # determine which unit is the primary
+    expected_primary = None
+    for unit in ops_test.model.applications[APP_NAME].units:
+        # connect to mongod
+        client = MongoClient(unit_uri(unit.public_address))
+
+        # check primary status
+        if client.is_primary:
+            expected_primary = unit.name
+
+    # verify that there is a primary
+    assert expected_primary
+
+    # check if get-primary returns the correct primary unit regardless of
+    # which unit the action is run on
+    for unit in ops_test.model.applications[APP_NAME].units:
+        # use get-primary action to find primary
+        action = await unit.run_action("get-primary")
+        action = await action.wait()
+        identified_primary = action.results["replica-set-primary"]
+
+        # assert get-primary returned the right primary
+        assert identified_primary == expected_primary
```

Comment:
I believe this can trigger an error if I correctly understood the action code in the comment about it returning the primary just when called in the primary unit.

Reply:
As the comment above, I confused the flow of the code.